### ISSUE TITLE
Allow multiple JSON roots in Puppeteer response

### ISSUE THREAD

**Issue by warpech** [Starcounter/Starcounter#4017]
For the starts, I am just signalling the problem that me and @tomalec have discussed yesterday.

Right now we have a feature that the first JSON response returned from a handler gets sent to client. Responses from middleware or mapping are ignored. You can affect this by setting the `PublicViewModel` property, but this does not solve blending at the root level -- still, there can only be one root.

This has numerous inelegant consequences:

- Puppeteer/PuppetJs session state version number (`_ver#c$`, `_ver#s`) is added as a "magic" root property. This alters the original JSON tree proposed by the app.
- It is not possible to get responses from multiple apps to a root request
- LauncherLayoutInfo (definition of the blending composition) cannot be added at the root level, meaning that the root partial is not blendable

Idea: Putting all Puppeteer responses in Starcounter to some JSON envelope solves all of the above problems:

```js
  {
    "_ver#c$": 0,
    "_ver#s": 0,
    "Responses": {
        /* All returned responses go here */
    }
  }
```

Solving this has many benefits:

- removes magic of version numbers
- turns the root response into just a partial. "Partials all the way up" AKA "Make Starcounter partial-centric" AKA https://github.com/Starcounter/Starcounter/issues/3540
- allows for cleaner implementation of many points in https://github.com/StarcounterPrefabs/Launcher/issues/261
- allows to add siblings a the root level, so common UI such as Layout Editor or Sign In can be used with any app through middleware or URI mapping
- Layout Editor will be finally able to edit root partials

I think this should be a goal for Starcounter 2.4.

Anyone has strong ideas for or against?

cc @chrhol, @per-samuelsson, @miyconst, @tomalec, @Starcounter-Jack

**Comment by chrhol** [comment]
I agree on these points. I don't really remember why we have special handling for the root. 

However, why use a different pattern with `Responses` here instead of just using the same pattern as we do in all other places for partials (i.e. currently using the name of the app)?

```
{
  "_ver#c$": 0,
  "_ver#s": 0,
  "Launcher": {
     
  },
  "SomeOtherApp": { 
    ...
  }
}
```

This would mean that we have no special treatment (other than for the versionnumbers, but that already exists) and the change for this on the server will be minimal. More or less removing an if-statement that checks if the current viewmodel is a root (= no parent) and add a line when evaluating an incoming patch that gets the correct viewmodel.

**Comment by warpech** [comment]
Good point, I am all for unifying the treatment of the root and the other places.

**Comment by chrhol** [comment]
I started implementing the changes and got stuck on additional things that would be nice to fix when doing this :smile: 

To change so that multiple viewmodels are allowed for the root is not a major thing, but to get rid of `PublicViewModel` more changes are needed on how patches and sessions work together. Maybe we need to separate public/exposed viewmodels and viewmodels stored for private use. We discussed that a long time ago but never got around to implement it. 

Something like this (but maybe with better names for the properties...), both public and private data is only visible for the app that added it, but public data will be sent to client.

```csharp
Session s = new Session();
s.PublicData = master;
s.PrivateData = someJson;
```

Then the data for all apps that is stored as `PublicData` will be sent to the client. The downside with this is that now we have two ways for the root to end up on the client, either by setting this property or having a mapping so that it's merged.

**Comment by chrhol** [comment]
Actually by changing to the sample above, it's makes more sense to follow the first example from @warpech. Having the viewmodels for the root as an array.

Then we support both apps responding with their own root viewmodel (that is NOT merged) or merged on the root depending on the need for the specific apps.

**Comment by warpech** [comment]
Again, thanks for picking this up! This will be immediately useful.

> Actually by changing to the sample above, it's makes more sense to follow the first example from @warpech. Having the viewmodels for the root as an array.

Actually in my example the viewmodels are a namespaced `Responses` object, not an array, but I guess this does not matter here.

If I understand correctly, what we have right now corresponds to `PublicData` in your example. I think that `PrivateData` is achievable already by:

- a `PublicData` class member that is not exposed in JSON
- by using `ConditionalWeakTable<MasterPage, MasterPagePrivateInfo>`

It seems to me that implementing `PrivateData` it is redundant then. Unless there is some additional benefit? Do we have some explicit requests for it?

**Comment by chrhol** [comment]
> It seems to me that implementing PrivateData it is redundant then. Unless there is some additional benefit? Do we have some explicit requests for it?

One thing I can think of would be, in the future, when the viewmodels will be semi-persistent and can survive a restart of the codehost. Then the private data (and all data that is cached and not stored as a property in json or in a database class) will be lost and would be needed to be recreated somehow.

How codehost restarts would work is kind of uncharted territory and I know we have done several things (both inside Starcounter and in apps) where this will not work. But maybe ultimately we want to revisit the whole concept of Session and viewmodels and how they work together.

**Comment by warpech** [comment]
Then maybe let's follow the [YAGNI](https://en.wikipedia.org/wiki/You_aren't_gonna_need_it) principle regarding `PrivateData`, meaning: let's get back to it when it's needed, not as part of this issue?

**Comment by chrhol** [comment]
> Then maybe let's follow the YAGNI principle regarding PrivateData, meaning: let's get back to it when it's needed, not as part of this issue?

Yes, good point. I will continue with doing the first steps first :smile:

**Comment by warpech** [comment]
Any progress on this one? It allows us to remove a nasty hack in CompositionEditor: https://github.com/Starcounter/CompositionEditor/issues/8

**Comment by warpech** [comment]
The partial returned as the main "root" should pull other partials mapped to "root" token using  `Blender.MapUri`.
